lib/src/import_path_parser.dart

```diff
@@ -0,0 +1,190 @@
+// Copyright (c) 2020, Google Inc. Please see the AUTHORS file for details.
+// All rights reserved. Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Based on the original work of:
+// - Jacob MacDonald: jakemac53 on GitHub
+//
+// Faster parser by:
+// - Graciliano M. Passos: gmpassos @ GitHub
+//
+
+import 'dart:io';
+
+import 'package:analyzer/dart/analysis/results.dart';
+import 'package:analyzer/dart/analysis/utilities.dart';
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:package_config/package_config.dart';
+import 'package:path/path.dart' as p;
+
+import 'import_path_base.dart';
+
+/// Dart import parser.
+/// Parses for import directives in Dart files.
+class ImportParser extends ImportWidget {
```

Comment:
I can't see any difference using the faster parser - and it's a lot of extra stuff to maintain. Ultimately, its only going to be a constant performance factor improvement and we are dealing with something exponential in its complexity, so a constant factor doesn't help. This implementation still looks at the whole file anyways though.

So I think we should drop the fast parser mode (or if you want, you could make it injectable or something and have your own implementation?). I don't want to support this parser ultimately, unless it made a really large difference in performance.

Reply:
Testing with 2 real projects, the difference with the faster parser is significant. It's 3x faster.

Reply:
That makes sense that its more noticeable in a real project, I realized that my pet project libraries only have directives 🤣 .

But, while it might be faster at parsing, that isn't overall the slow thing here, and I really don't want to own/maintain a regex based parser.

If we just allow this class to accept an ImportParser (which should really only do String -> dependencies conversion), then you could own this regex based one and use it in your own scenarios?

Reply:
Well, it's a very simple `RegExp`. The code finds the last import, then there's a fallback to the full parser in case of error. The idea is not to give a perfect fast parser, just an alternative.

We are using only the fast parser here for out projects. (already using the tool here).

Reply:
FYI:

https://github.com/dart-lang/sdk/issues/53498

Reply:
> Well, it's a very simple `RegExp`. The code finds the last import, then there's a fallback to the full parser in case of error. The idea is not to give a perfect fast parser, just an alternative.
> 
> We are using only the fast parser here for out projects. (already using the tool here).

But it will still have to be updated as the language evolves (we have multiple proposals out currently that could affect this, they may not land, but I don't have a lot of time to spend maintaining this).

> FYI:
> 
> [dart-lang/sdk#53498](https://github.com/dart-lang/sdk/issues/53498)

I definitely think it would be great if we could get some better support from the SDK here, we do this in a lot of places.